src/sinks/redis.rs

```diff
@@ -0,0 +1,485 @@
+        };
+
+        let key = self.key.render_string(&event).map_err(|error| {
+            emit!(TemplateRenderingFailed {
+                error,
+                field: Some("key"),
+                drop_event: true,
+            });
+        })?;
+
+        let encoded = encode_event(event, &self.encoding);
+        let message_len = encoded.len();
+
+        let data_type = self.data_type;
+        let method = self.method;
+
+        self.state = RedisSinkState::Sending(Box::pin(async move {
+            let send = match (data_type, method) {
+                (DataType::List, Method::LPush) => ConnectionManager::lpush,
+                (DataType::List, Method::RPush) => ConnectionManager::rpush,
+                (DataType::Channel, _) => ConnectionManager::publish,
+            };
+
+            let result = send(&mut conn, key.clone(), encoded.clone()).await;
+            if result.is_ok() {
+                emit!(RedisEventSent {
+                    byte_size: message_len,
+                });
+            }
+
+            (conn, result)
+        }));
+
+        Ok(())
+    }
+
+    fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        ready!(self.poll_in_flight_prepare(cx));
+
+        let this = Pin::into_inner(self);
+        while !this.in_flight.is_empty() {
+            match ready!(Pin::new(&mut this.in_flight).poll_next(cx)) {
+                Some((seqno, Ok(result))) => {
+                    trace!(
+                        message = "Redis sink produced message.",
+                        length = %result,
+                    );
+                    this.pending_acks.insert(seqno);
+                    let mut num_to_ack = 0;
+                    while this.pending_acks.remove(&this.seq_tail) {
+                        num_to_ack += 1;
+                        this.seq_tail += 1
+                    }
+                    this.acker.ack(num_to_ack);
+                }
+                Some((_, Err(error))) => {
+                    error!(message = "Redis sink generated an error.", %error);
+                    emit!(RedisSendEventFailed { error });
+                    return Poll::Ready(Err(()));
+                }
+                None => break,
+            }
+        }
+        Poll::Ready(Ok(()))
+    }
+
+    fn poll_close(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.poll_flush(cx)
+    }
+}
+
+async fn healthcheck(mut conn: ConnectionManager) -> crate::Result<()> {
+    redis::cmd("PING")
+        .query_async(&mut conn)
+        .await
+        .map_err(Into::into)
+}
+
+fn encode_event(mut event: Event, encoding: &EncodingConfig<Encoding>) -> String {
+    encoding.apply_rules(&mut event);
+
+    match encoding.codec() {
+        Encoding::Json => serde_json::to_string(event.as_log())
+            .map_err(|error| emit!(RedisEncodeEventFailed { error }))
+            .unwrap_or_default(),
```

Comment:
I think you can just `expect` here rather than emitting an event when the event fails to encode. It should never fail to encode in practice. You could then drop that internal event. See `http` sink for example: https://github.com/timberio/vector/blob/29736f34150da640175dae4aeb4ff8638cd1c441/src/sinks/http.rs#L206-L208